src/app/core/models/undoredo.model.ts

```diff
@@ -0,0 +1,53 @@
+export class UndoRedo<T> {
+    historyArray: Array<T>;
+    currIndex: number;
+    firstEntryIndex: number;
+    lastEntryIndex: number;
+
+    constructor(capacity: number, initialValue: T) {
+      this.historyArray = new Array<T>(capacity);
+      this.firstEntryIndex = 0;
+      this.lastEntryIndex = 0;
+      this.currIndex = 0;
+      for (let i = 0; i < capacity; i++) {
+        // initialize value
+        this.historyArray[i] = initialValue;
+      }
+    }
+
+    updateEntry(entry: T): void {
+      this.currIndex = this.incrementIndex(this.currIndex);
+      this.lastEntryIndex = this.currIndex; // resets future history
+      if (this.currIndex === this.firstEntryIndex) {
```

Comment:
Could we handle this case more elegantly? e.g. notify users there is a limit instead of overwriting the history

Reply:
I think for undo and redo, notifying the user explicitly seems a bit overkill and I believe users would have experiences with text editors where there is a limit to how much history the application can store so I think this shouldn't really be an issue that we need to explicitly tell the users. Since the users would realise that they are unable to undo till a certain point. I think a better alternative would just be incrementing the initial size of the array or changing when the text is being stored.

However, I can indeed change this to use a dynamic array instead of a circular array which will make the insertion amortized O(1) + time taken to copy the text. A potential downside will be the rare possibility of lag when there is a very large history stack and the insertion of a new element caused the array to resize (which can be fixed with a async solution).

Reply:
line 20 here resets the future history.
Basically when a user undo to a previous state and make changes on that state, the future state should be reset.
e.g.
`[ 1,2,3,4,*5* ] -undo-> [ 1,*2*,3,4,5 ] -redo-> [1,2,*3*,4,5] -modify-> [1,2,3,*3.5*]`
when the user makes an edit on state 3 to state 3.5. State 4 and 5 should no longer be accessible hence the resetting future history.